lib/products.go

```diff
@@ -0,0 +1,151 @@
+package lib
+
+import "fmt"
+
+const legacyProductId = "terraform"
+
+type ProductDetails struct {
+	ID                    string
+	Name                  string
+	DefaultMirror         string
+	VersionPrefix         string
+	DefaultDownloadMirror string
+	ExecutableName        string
+	ArchivePrefix         string
+	PublicKeyId           string
+	PublicKeyUrl          string
+}
+
+type TerraformProduct struct {
+	ProductDetails
+}
+type OpenTofuProduct struct {
+	ProductDetails
+}
+
+type Product interface {
+	GetId() string
+	GetName() string
+	GetDefaultMirrorUrl() string
+	GetVersionPrefix() string
+	GetExecutableName() string
+	GetArchivePrefix() string
+	GetPublicKeyId() string
+	GetPublicKeyUrl() string
+	GetShaSignatureSuffix() string
+	GetArtifactUrl(mirrorURL string, version string) string
+}
+
+// Terraform Product
+func (p TerraformProduct) GetId() string {
+	return p.ID
+}
+func (p TerraformProduct) GetName() string {
+	return p.Name
+}
+func (p TerraformProduct) GetDefaultMirrorUrl() string {
+	return p.DefaultMirror
+}
+func (p TerraformProduct) GetVersionPrefix() string {
+	return p.VersionPrefix
+}
+func (p TerraformProduct) GetExecutableName() string {
+	return p.ExecutableName
+}
+func (p TerraformProduct) GetArchivePrefix() string {
+	return p.ArchivePrefix
+}
+func (p TerraformProduct) GetArtifactUrl(mirrorURL string, version string) string {
+	return fmt.Sprintf("%s%s", mirrorURL, version)
```

Comment:
This assumes `mirrorURL` is always suffixed with `/`, right?
Would it make sense to 1) trim trailing slash (if any) and 2) use `%s/%s` instead to ensure we glue URL parts together correctly? 🤔

Reply:
Without looking (I'll take another look at the code base tomorrow morning), I saw we always prepend the trailing slash - I had at some point had the seperator here, which broke the requests because of double slashes.
I wasn't keen on it at the time, so maybe it's better to check when the URL is being used.
I'll look at cleaning this up

Reply:
Thanks.
This is actually what I try to implement always for this kind of _base_ URL values: trim trailing slash if any and append slash unconditionally then — this way I'm always sure there's no doubleslash at this place.

Reply:
This is now fixed :)